Enhance caching

SailReal opened the pull request:
This PR enhances the caching behavior of this library in the following way:

* Storing the already visited remote file hierarchy in a local tree structure and executing future requests against this local tree, including updating against local and cloud-side changes
* Increased default FileHeader expiration from 1 --> 5 seconds to not request the header on each GET request
* Requests for the FileHeader and DirectoryID cache are now not executed several times simultaneously

SailReal committed:
WIP enhance caching to reduce amount of queries against the cloud

Use a tree node structure to store the PROPFIND-entries. Currently the tree is only created while startup and doesn't get refreshed after a specific amount of time or when there are changes made by the client. However, it can still be used to test its benefits.

SailReal committed:
Use generic tree node structure

overheadhunter committed:
Merge branch 'develop' into feature/enhance-caching

overheadhunter committed:
Added some tests for WebDavTreeNode

overheadhunter committed:
Renamed WebDavTreeNode → CachedNode and added NodeCache for complex operations on cached nodes

overheadhunter committed:
Implemented NodeCache delete

overheadhunter committed:
Implemented NodeCache.delete

overheadhunter committed:
Removed recursive marking of dirty nodes fro CachedNode class

overheadhunter committed:
implemented NodeCache.markDirty

SailReal committed:
Partially use NodeCache in WebDavClient 

Still some TODOs like e.g.

- [ ] Implement add to cache
- [ ] Handle HTTP status code 412
- [ ] Handle HTTP status code 404 (delete from cache?)
- [ ] Enhance String/Path handling (fullPath)
- [ ] Implement tests
- [ ] Delete MetadataCachingProviderDecorator

SailReal committed:
Use of NodeCache in WebDavClient improved, still WIP

SailReal committed:
Check whether caching is supported on the server side

SailReal committed:
Use method instead of function, as @overheadhunter suggested

overheadhunter committed:
Cache improvements

overheadhunter committed:
more tests

overheadhunter committed:
fixed move of uncached node

overheadhunter committed:
return 404 responses from cache

SailReal committed:
Fix integration tests

SailReal committed:
Fix PropfindItem handling

SailReal committed:
Add further tests, still some pending in CachedPropfindEntryProviderTest

SailReal committed:
Fix tests

SailReal committed:
Implement further CachedPropfindEntryProvider tests

overheadhunter committed:
use only one dedicated constructor

SailReal committed:
Fix move if the name of the node changed

SailReal committed:
Fix move if the name of the node changed 2.0

Also consider the descendants of n

SailReal committed:
Fix dirty flag removed when dirty child is added with existing data

As well as remove constructor which is only used for test purposes

SailReal committed:
Reanimate itemMetadataCache to respond to spam-requests only once

We should think about if this is cloud-agnostic and is therefore the correct layer or it should be moved into the cloud-specific part

SailReal committed:
Implement further test

SailReal committed:
Don't add children to new parent as the name depend on the parent DirID

This works only if we have a 1:1 mapping of source and target folder that means move(/foo/bar, /foo/baz) which is not the case using our CryptoDecorator, there the cipher text name depend on the parent dir so we cannot simply append the child to the new parent. At this layer we can only mark the target parent as dirty so it will be fetched from the cloud.

SailReal committed:
Marking children already fetched to avoid mistakenly empty folders

SailReal committed:
When itemMetadata is called using the cache execute PROPFIND DEPTH=1

To reduce the amount of requests, itemMetadata's delegated to the cache but executed against the cloud uses DEPTH=1 to get the children right along with them.

SailReal committed:
Revert "Don't add children to new parent as the name depend on the parent DirID"

This reverts commit c6935437fba0dfb4e2758e012b37454b655d8d02.

SailReal committed:
Return target node not source from NodeCache after move

SailReal committed:
Report folders as empty only when children already fetched

And cached node is empty

SailReal committed:
If parent not in cache, request it in CachedPropfind's itemMetadata()

If parent is not available in cache, request parent in CachedPropfindEntryProvider's itemMetadata to avoid listings against `/foo/bar/baz` while `/foo/bar/` has not yet been queried which can lead to spam requests as `isChildrenFetched` is in this case not set in the parent and would also not be set using a PROPFIND on the child. Therefore listing on the parent.

SailReal committed:
Move NodeCache and CachedNode to the correct package

SailReal committed:
Add response of a PUT into the cache and aggregate list-requests too

SailReal committed:
WIP implement RemoteChangeDetector

SailReal committed:
Enhance and harden the RemoteChangeDetector

* Call `updateChildren(node)` also for folders without cached data as successors may still need to be updated, the termination condition is ETag present and not changed in the respective subtree. 
* Set `setChildrenFetched` for the current folder so that further requests are executed completely against the cache if updated using the remote change detector
* Catch `Exception` during the execution of the remote change detector so that it is re-triggered. Without catching exceptions, no further execution.

SailReal committed:
Protecting children from concurrent modifications that may occur

SailReal committed:
Request FileHeader and DirectoryID once, even if response isn't arrived

Before it was possible that the same DirectoryID or FileHeader was requested multiple times if the response wasn't available yet while a caller asks for it (again).

SailReal committed:
Increase default FileHeader cache expiration from 1 to 5 seconds

SailReal committed:
Cleanup code

overheadhunter committed:
inlined FileHeaderCache

SailReal committed:
Expire header cache entries after write and not after access

SailReal committed:
Fix WebDAV integration tests

SailReal committed:
Remove synchronized as the Guava cache is threadsafe

As well as  `cache.invalidate(nonExistingKey)` is a noop

SailReal committed:
Apply suggestions of Codacy Static Code Analysis

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java:
```diff
@@ -15,38 +14,43 @@
 	private final static int DEFAULT_CACHE_TIMEOUT_SECONDS = 10;
 
-	final Cache<CloudPath, CompletionStage<CloudItemMetadata>> itemMetadataCache;
+	final Cache<CloudPath, CompletionStage<CloudItemMetadata>> itemMetadataRequestAggregator;
+	final Cache<Map.Entry<CloudPath, Optional<String>>, CompletionStage<CloudItemList>> cloudItemListRequestAggregator;
```
I'd rename these caches, as "aggregator" suggest some aggregation logic.

```suggestion
	final Cache<CloudPath, CompletionStage<CloudItemMetadata>> cachedItemMetadataRequests;
	final Cache<Map.Entry<CloudPath, Optional<String>>, CompletionStage<CloudItemList>> cachedItemListRequests;
```

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java:
```diff
@@ -15,38 +14,43 @@
+	public MetadataRequestAggregator(CloudProvider delegate, Duration cacheEntryMaxAge) {
 		this.delegate = delegate;
-		this.itemMetadataCache = CacheBuilder.newBuilder().expireAfterWrite(cacheEntryMaxAge).build();
+		this.itemMetadataRequestAggregator = CacheBuilder.newBuilder().expireAfterWrite(Duration.ofSeconds(1)).build();
+		this.cloudItemListRequestAggregator = CacheBuilder.newBuilder().expireAfterWrite(Duration.ofSeconds(1)).build();
```
Duration should be configurable

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java:
```diff
@@ -70,22 +74,24 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-						cloudItemList.getItems().forEach(metadata -> itemMetadataCache.put(metadata.getPath(), CompletableFuture.completedFuture(metadata)));
-					}
-				});
+		try {
+			return cloudItemListRequestAggregator.get(Map.entry(folder, pageToken), () -> delegate.list(folder, pageToken).whenComplete((metadata, throwable) -> {
+				// immediately invalidate cache in case of exceptions (except for NOT FOUND):
+				if (throwable != null && !(throwable instanceof NotFoundException)) {
+					invalidateAggregators(folder);
+				}
+			}));
+		} catch (ExecutionException e) {
+			return CompletableFuture.failedFuture(e);
+		}
```
1. use `cache.getUnchecked()` to avoid try-catch (only if we know for sure that the loading method doesn't throw checked exceptions)
2. instead of two nested lambdas, maybe the outer lambda can be a `private CompletionStage<CloudItemList> listUncached(folder, pageToken) {...}`

overheadhunter replied:
Same for `itemMetadata(...)` method btw :wink:

SailReal replied:
Only the `LoadingCache` has a `getUnchecked` method,

overheadhunter replied:
I see, then proceed to point 2 🙂

SailReal replied:
Still WIP :wink:

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java:
```diff
@@ -95,68 +101,63 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
+	private void invalidateAggregators(CloudPath cleartextPath) {
+		itemMetadataRequestAggregator.invalidate(cleartextPath);
+
+		for (var path : cloudItemListRequestAggregator.asMap().keySet()) {
+			if (path.getKey().equals(cleartextPath)) {
+				cloudItemListRequestAggregator.invalidate(path);
+				return;
 			}
 		}
```
what is this loop supposed to do? looks to me as if it only invalidates the first key that matches the given path. why not just use `invalidate(cleartextPath)` directly?

SailReal replied:
The return was a faulty optimization. As the key of the `cloudItemListRequestAggregator` is a Map containing the `Path` and `Pagetoken`, we need to iterate over all Paths and remove those entries matching this path.

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/webdav/CachedPropfindEntryProvider.java:
```diff
@@ -0,0 +1,230 @@
+		this.rootPoller = rootPoller;
+		this.cacheUpdater = cacheUpdater;
+
+		var executor = Executors.newSingleThreadScheduledExecutor();
+		var remoteChangeDetector = new RemoteChangeDetector();
+		executor.scheduleWithFixedDelay(remoteChangeDetector, config.getRemoteChangePollerInitialDelaySeconds(), config.getRemoteChangePollerPeriodSeconds(), TimeUnit.SECONDS);
```
despite the fact that no scheduler is required (we only need to expose the `pollRemoteChanges()` method), it is always good to keep a strong reference to an executor, so we can gracefully shut it down later.

SailReal replied:
I just moved the `TimerTask` into the `WebDavIntegrationTest`

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/webdav/CachedPropfindEntryProvider.java:
```diff
@@ -0,0 +1,230 @@
+
+	public PropfindEntryItemData itemMetadata(CloudPath path, Function<CloudPath, List<PropfindEntryItemData>> parentLoader, Function<CloudPath, List<PropfindEntryItemData>> pathLoader) {
+		var cachedNode = cache.getCachedNode(path);
+		Optional<CachedNode> cachedParent = path.getParent() != null ? cache.getCachedNode(path.getParent()) : Optional.empty();
```
```suggestion
		Optional<CachedNode> cachedParent = Optional.ofNullable(path.getParent()).flatMap(cache::getCachedNode);
```

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/webdav/CachedPropfindEntryProvider.java:
```diff
@@ -0,0 +1,230 @@
+
+	public List<PropfindEntryItemData> list(CloudPath path, Function<CloudPath, List<PropfindEntryItemData>> loader) throws CloudProviderException {
+		var cachedNode = cache.getCachedNode(path);
+		Optional<CachedNode> cachedParent = path.getParent() != null ? cache.getCachedNode(path.getParent()) : Optional.empty();
```
```suggestion
		Optional<CachedNode> cachedParent = Optional.ofNullable(path.getParent()).flatMap(cache::getCachedNode);
```

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/webdav/CachedPropfindEntryProvider.java:
```diff
@@ -0,0 +1,230 @@
+				// if local exists and same ETAG --> ignore further sub-tree ✓
+				// if remote exists but not local --> create ✓
+
+				for (CachedNode localChild : localChildren) {
+					var remoteItemMetadata = remoteChildren.stream().filter(remote -> remote.getName().equals(localChild.getName())).findAny();
+					if (remoteItemMetadata.isPresent()) {
+						var localData = localChild.getData(PropfindEntryItemData.class);
+						if (localData != null) {
+							if (!localData.isSameVersion(remoteItemMetadata.get())) {
+								cache.getCachedNode(node.resolve(localChild.getName())).get().update(remoteItemMetadata.get());
+								if (localData.isCollection()) {
+									updateChildren(node.resolve(localChild.getName()));
+								}
+							} // else branch does nothing, even if a subtree exists but ETag of the parent didn't change
+						} else {
+							cache.getCachedNode(node.resolve(localChild.getName())).get().update(remoteItemMetadata.get());
+							if(remoteItemMetadata.get().isCollection()) {
+								updateChildren(node.resolve(localChild.getName()));
+							}
+						}
+					} else {
+						cache.delete(node.resolve(localChild.getName()));
+					}
+				}
```
high complexity, could use some private subfunctions

SailReal replied:
Still not good but with https://github.com/cryptomator/cloud-access-java/pull/5/commits/bba82105a174f910fa8d848ea36978fadfd3d69b maybe a little bit better

overheadhunter replied:
sufficient for now 👍

overheadhunter reviewed (changes requested):
To be honest, I didn't completely check the WebDAVClient itself, but found some room for simplification in the cache-related classes

SailReal committed:
Apply suggestions from code review

Co-authored-by: Sebastian Stenzel <overheadhunter@users.noreply.github.com>

SailReal committed:
Remove early return as we need to loop over all paths and remove them

SailReal committed:
Finish rename of cachedItemMetadataRequests and cachedItemListRequests

SailReal committed:
Make max age of requests aggregation configurable

SailReal committed:
RemoteChangeDetector divided into several methods, can still be improved

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java:
```diff
@@ -32,15 +33,17 @@
+		this(
+				delegate, //
+				Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds", DEFAULT_CACHE_TIMEOUT_SECONDS)), //
+				Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.aggregateRequestsMaxAgeSeconds", DEFAULT_AGGREGATE_REQUESTS_MAX_AGE)) //
```
In my opinion, this is a misleading configuration name. If I'm not mistaken, the purpose of this class is _not only_ the **new** feature to avoid duplicate requests while a request is still running, but still to cache metadata responses (and now also list responses).

Even if we received a response, it makes sense to keep the completed future in our cache. After all the purpose of this caching layer is to not only avoid request while waiting for a response, but also a few seconds after.

So my suggestion is:
1. Either use `org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds` for all three caches.
2. Or use three separate config properties, i.e. `metadataTimeoutSeconds`, `itemListTimeoutSeconds`, `quotaTimeoutSeconds`

SailReal replied:
> If I'm not mistaken, the purpose of this class is not only the new feature to avoid duplicate requests while a request is still running, but still to cache metadata responses (and now also list responses).
> 
> Even if we received a response, it makes sense to keep the completed future in our cache. After all the purpose of this caching layer is to not only avoid request while waiting for a response, but also a few seconds after.

We now use the `MetadataRequestAggregator` regarding `itemMetadata` and `list` only for aggregating responses so that a list request which is already running not executed a second and a "thousand" time as these methods are often called very quickly one after the other on the same path but as soon as the response is available, the caching is handled by our `NodeCache` in the `CachedPropfindEntryProvider`. The `NodeCache` knows due to `ETag` the best if a cached entry is still valid that is why the response in the `MetadataRequestAggregator` should be in my opinion invalidated as soon as possible so that we have only one cache and one aggregator running and not two caches.

overheadhunter replied:
Oh ok I see. While this is capable of cashing, it is not intended to act as a cache (at least not if the backend supports etags).

So the question is if we want it to show this capability or not. If not, we should move the quota thing to a different place, shouldn't we?

SailReal replied:
We can go even further and remove the expiration from the `cachedItemMetadataRequests` and `cachedItemListRequests` completely but invalidate it as soon as the response is available, IMO this is the correctest version, e.g. `itemMetadata` would look like this:

```java
return cachedItemMetadataRequests.get(node, () -> delegate.itemMetadata(node)
			.whenComplete((metadata, throwable) -> invalidateAggregators(node)));
```
Then we have no overlapping if e.g. the response isn't fast enough and the functionality of it is clearer.

---

Regarding the quota cache I totally agree with you, quota is somehow wrong here.

---

I think we should
1. As mentioned above, remove the expiration from `cachedItemMetadataRequests` and `cachedItemListRequests` and invalidate immediately after receiving the response
2. Create a new `QuotaCachingProviderDecorator` which caches `quota` requests
3. Re-Implement the previous functionality of the `MetadataCachingProviderDecorator` in case of a cloud doesn't support caching like WebDAV without ETags so that the cloud state is present after a request until the entry gets invalidated e.g. due to the timeout. But use this inside of e.g. the `WebDAVProvider` as we don't know from the outside if we can use caching.

overheadhunter replied:
Moving from cache to a simple map + manual removal was my first thought, too.

My second thought, though, was: Would it hurt to keep this as a second caching layer? Thinking of servers without etag support here.

Edit: I should have read your third point more carefully :D So yes, the current implementation is actually exactly what this `MetadataCachingProviderDecorator` should look like (*including* the quota cache, as long as timeouts are configurable separately!).

Bonus question: Can timeout be a magic value (e.g. `0`), which expires immediately but still prevents duplicate requests? This would remove the need for a separate request aggregator class.

SailReal replied:
`expireAfterWrite(0)` looks good:

> When duration is zero, this method hands off to maximumSize(0), ignoring any otherwise-specificed maximum size or weight. This can be useful in testing, or to disable caching temporarily without a code change. 

`maximumSize(0)` in return leads to

> ...elements will be evicted immediately after being loaded into cache. This can be useful in testing, or to disable caching temporarily without a code change. 

I'll implement it this way :+1:

SailReal replied:
Needs still some further unit tests.

ben-manes replied:
FYI, you might benefit from using [Caffeine](https://github.com/ben-manes/caffeine) here instead of Guava due to native support for futures via [AsyncCache](https://github.com/ben-manes/caffeine/wiki/Population#asynchronous-manual). An in-flight future will not be eligible for eviction by a maximum size or expiration policy, and will be automatically removed if the future fails or resolves to a `null` value. That might simplify your code, but I couldn't tell if you wanted to cache exceptions or not.

The comments about preventing duplicate requests with immediate eviction would in your Guava code mean only for the duration of creating the future (usually fast), whereas in Caffeine it would be until the future completed. I suspect the latter is what you'd want.

overheadhunter replied:
@ben-manes That does indeed look interesting (for various applications). We'll have a look at it, thanks for the hint!

SailReal replied:
>  you might benefit from using Caffeine here instead of Guava due to native support for futures via AsyncCache. 

Even Google states ["Prefer Caffeine over Guava's caching API"](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/CacheBuilder.html), thanks for your input @ben-manes, it looks really great and we can definitely use Caffeine in other places. Here I'm still evaluating, but I don't think so, because under certain circumstances (the cloud or a server doesn't support smarter caching) we actually want to prevent not only duplicates on this level but really cache, which includes exceptions.

> The comments about preventing duplicate requests with immediate eviction would in your Guava code mean only for the duration of creating the future (usually fast), whereas in Caffeine it would be until the future completed. I suspect the latter is what you'd want.

As we (currently) don't set an expiration duration but invalidate manually, this doesn't apply AFAIK (and tested it):

> These features are all optional; caches can be created using all or none of them. By default cache instances created by CacheBuilder will not perform any type of eviction. (from https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html)

overheadhunter reviewed (changes requested):
I'm still not a fan of the "aggregate requests" naming. Recently we focused on avoiding duplicate requests, but don't forget that the whole point of this class is to cache responses :wink:

SailReal committed:
Cache CloudImpl response if cannot cache itself,otherwise aggregate only

TODO write some further tests in MetadataCachingProviderDecoratorTest

SailReal committed:
Remove RemoteChangeDetector from the CachedPropfindEntryProvider

Now we can call from outside `pollRemoteChanges`

SailReal committed:
Fix build -.-

SailReal committed:
Remove unused parameter

SailReal committed:
Evict cache entries directly after loading finished

Later re-implement  response caching next to response aggregation if the CloudProviderImpl does not provide a better caching mechanism

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java:
```diff
@@ -191,9 +157,11 @@ private void evict(CloudPath cleartextPath) {
+
+	private void evict(Map.Entry<CloudPath, Optional<String>> folderKey) {
+		cachedItemListRequests.invalidate(folderKey);
+	}
```
bad method name (as it is unclear whether this evicts items from `cachedItemListRequests` or `cachedItemListRequests`

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java:
```diff
@@ -70,94 +70,98 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-					}
-				});
+		try {
+			var folderKey = Map.entry(folder, pageToken);
```
I'd suggest using a record. Misusing `Map.Entry` to fix Java's lack of tuples feels wrong.

overheadhunter replied:
And the type declarations for the Cache are way too long.

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java:
```diff
@@ -70,94 +70,98 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-					itemMetadataCache.put(file, future);
-					quotaCache.invalidateAll();
-				});
+		return delegate.deleteFile(file);
```
no longer putting 404 response into the cache? previous implementation looked more efficient

SailReal replied:
In the currently published version, as discussed, the cache is only used to remove duplicate requests (Request-Aggregation). In this case, the respective cloud implementation needs to handle better caching, which also includes 404.

In a subsequent commit, this cache is extended by a parameter that we can use to control whether only aggregation or real caching is used. In case of real caching the `expireAfterWrite` has to be set again and `whenComplete` looks a bit different (especially regarding this method, as you said, we will add again 404 responses but only if caching is required in this layer). Afterwards the cache handles both variants: Request-Aggregation and if necessary Caching.

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java:
```diff
@@ -70,94 +70,98 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-					itemMetadataCache.put(folder, future);
-					quotaCache.invalidateAll();
-				});
+		return delegate.deleteFolder(folder).whenComplete((nullReturn, exception) -> evictIncludingDescendants(folder));
```
see previous comment

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/webdav/WebDavClient.java:
```diff
@@ -399,9 +415,18 @@ void checkServerCompatibility() throws ServerNotWebdavCompatibleException {
+		return loadPropfindItem(CloudPath.of("/"));
+	}
+
+	void canUseCaching(PropfindEntryItemData propfindEntryItemData) throws UnauthorizedException {
```
changing this object's state from within a method called "canSomthing" feels like a side effect... Maybe we can find out capabilities from within a WebDavClientBuilder and whether etag caching is used or not be an immutable property set in the WebDavClients constructor?

overheadhunter commented on src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java:
```diff
@@ -24,31 +24,31 @@
 	public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEntryMaxAge) {
 		this.delegate = delegate;
-		this.itemMetadataCache = CacheBuilder.newBuilder().expireAfterWrite(cacheEntryMaxAge).build();
+		this.cachedItemMetadataRequests = CacheBuilder.newBuilder().build();
+		this.cachedItemListRequests = CacheBuilder.newBuilder().build();
```
If there is no expiration policy, there is no point in using Guava's Cache, I guess. Either use a normal ConcurrentHashMap or take a look at Caffeine.

SailReal replied:
I strongly assume that the variant of https://github.com/cryptomator/cloud-access-java/pull/5#discussion_r781257124 is easier and less DRY-code instead of implementing another decorator in case the cloud does not support specific caching and we want to cache the responses for a certain time.

SailReal committed:
Make CachedPropfindEntryProvider in WebDavClient final

SailReal committed:
Switch to Record instead of Map to fix Java's lack of tuples

Bump Java to version 17

SailReal committed:
Cache responses if the CloudImpl doesn't have caching capabilities

SailReal committed:
Rename evict methods

SailReal committed:
Update GitHub runner to Java 17

SailReal committed:
Update maven-javadoc-plugin too and remove unused import

SailReal committed:
Bump jacoco-maven-plugin version as well
